Update setup, config and all dependencies to our most current state

jackie-ak opened the pull request.

jackie-ak committed:
change: adopt our current docker setup approach

jackie-ak committed:
Squashed 'config/' changes from 379a1ec..c262021

c262021 Merge pull request #3 from base-angewandte/feature/update-pre-commit-hooks
77f0503 build(deps): update base pre-commit-hooks to 2.1
66e39e6 Merge pull request #2 from base-angewandte/feature/date-format
846065f chore: update .gitignore
f496028 feat: set LC_ALL and use date format to be more consistent
2f2689e feat: add compose.override.yaml to gitignore generation
2c35c38 change: ignore SIM105
715bbf9 change: correct per-file-ignores
202373e change: ignore S101 for tests
c7a476c chore: fix typo
628e897 fix: change rev to string
f3f889c Merge branch 'feature/init'
686789f Merge branch 'feature/uv'
7dd7804 Merge branch 'feature/ruff'
61a7731 chore: update pre-commit-hooks rev to 2.0
30a25d6 change: exchange pip-tools with uv
823b7bc feat: add uv config
5775407 feat: add init-dev
ea54f18 change: if DEBUG=True run pre-commit-init on init
70224a5 feat: add ruff configuration and remove deprecated configuration
f8430bb refactor: move commitizen configuration to .cz.toml
6e80622 feat: add makemessages-docker and compilemessages-docker commands
2ccc1eb chore: change docker-compose to docker compose
b711234 docs: add README
326d08a feat: add .gitignore created from make-gitignore.sh
6287abf feat: add base ignore section

git-subtree-dir: config
git-subtree-split: c2620215fba590815df108619f3face15069b879

jackie-ak committed:
Merge commit '80f52ee13def88be23948fcfb698983dd78e3625' into feature/refactor-setup

jackie-ak committed:
change: update config

jackie-ak committed:
change!: consolidate config to a single .env file

jackie-ak committed:
docs: add notes on config file changes

jackie-ak committed:
change: .env settings for migration script

jackie-ak committed:
change: add sentry and healthchecks, adopt uv-based setup

jackie-ak committed:
docs: adapt docs to new setup

jackie-ak committed:
refactor: move process_user_attributes to accounts

jackie-ak committed:
refactor: use base_common instead of general

jackie-ak committed:
build: update to python 3.12

jackie-ak committed:
build(deps): update all packages

jackie-ak committed:
fix: use explicit serializers for autosuggest views

jackie-ak committed:
perf: use list comprehension for user list

jackie-ak committed:
Merge commit 'db2710fcff373d66f4fccfe0795f5c70a9fdb5b7' into feature/refactor-setup

jackie-ak committed:
Squashed 'config/' changes from c262021..eccc8bd

eccc8bd Merge branch 'feature/new-setup'
83dc1eb remove: remove pre-commit-init from init and init-dev
d0f9251 fix: sync with requirements-dev.txt in development setups
067ec6d fix: fix compile commands by removing src/
d96cfa8 change: update pre-commit commands to use uv

git-subtree-dir: config
git-subtree-split: eccc8bd59799cd65b265e9a2f8fd9ec09b665cd9

pmayer commented on compose.yaml:
```diff
@@ -8,29 +7,39 @@ services:
-      - POSTGRES_PASSWORD=$BASEAUTH_DB_PASSWORD
-      - POSTGRES_USER=$BASEAUTH_DB_USER
-      - POSTGRES_DB=$BASEAUTH_DB_NAME
+      - POSTGRES_PASSWORD=$POSTGRES_DB
```
The db and password variables here are wrong, it should be:

```
- POSTGRES_DB=$POSTGRES_DB
- POSTGRES_USER=$POSTGRES_USER
- POSTGRES_PASSWORD=$POSTGRES_PASSWORD
```

jackie-ak replied:
Oh wow, indeed, how did that even work? I guess because it is only assigned once on db creation. Never mind, fixed it with a commit.

pmayer commented on docker/cron/crontab:
```diff
@@ -1 +1,2 @@
+* * * * * docker container ls -f health=unhealthy -f name=portfolio --format "{{.Names}}" | xargs --no-run-if-empty docker restart | xargs -I{} sh -c '[ -n "${SENTRY_DSN}" ] && sentry-cli send-event -m "Unhealthy docker container {} restarted" -E ${SENTRY_ENVIRONMENT}'
```
here `name=portfolio` should be `name=baseauth`

pmayer commented on docs/source/install.md:
````diff
@@ -24,41 +24,53 @@ There are two supported ways to start the development server:
-  cp docker-compose.override.dev.yml docker-compose.override.yml
+  cp compose.override.dev.yaml compose.override.yaml
   ```
 
````
Should we add the recommendation to install the pre-commit hooks here?

````
- Install pre-commit hooks:

  ```bash
  pre-commit install
  ```
````

jackie-ak replied:
Yes, but I added it to the git checkout, because it feels more connected to that, instead of doing it after the .env and compose override files:

- Clone git repository, checkout branch `develop` and install pre-commit hooks:

  ```bash
  git clone https://github.com/base-angewandte/baseauth.git
  cd baseauth
  git checkout develop
  pre-commit install --install-hooks --overwrite
  ```

pmayer commented on docs/source/install.md:
````diff
@@ -139,8 +136,6 @@ automagically. Take a look at the [](./configuration.md) section for details.
   sudo apt-get update
   ```
 
-- Install docker and docker-compose
````
why did you completely remove that line? shouldn't we keep it as `- Install docker with compose plugin`

jackie-ak replied:
Good question, not sure tbh.
We already mention Docker in the requirements, but as we also explicitly have it as a step for the dev setup I added it again also to the prod setup.

pmayer commented on docs/source/requirements.md:
```diff
@@ -14,4 +18,5 @@ If you want to actively develop or run a development instance directly on your
 host (aka _the full developer setup_), you will also need:
 
 - Dev libraries for: `libldap2-dev libsasl2-dev libssl-dev`
```
do we need these libraries also with uv?

jackie-ak replied:
Not sure, I would assume so. But will test this later. Leaving it unsesolved for now.

jackie-ak replied:
Seems to work fine without them on my machine. But will test it on a fresh setup in a VM, just to be sure.

jackie-ak replied:
Ok, turns out, for local dev we still need those libraries. But I removed libssl-dev, as this is already listed in our general local dev setup guide, because it is also needed in other projects. LDAP and SASL are specific for baseauth.

jackie-ak committed:
chore: apply linting changes to full codebase

pmayer commented on docs/source/requirements.md:
```diff
@@ -14,4 +18,5 @@ If you want to actively develop or run a development instance directly on your
 
 - Dev libraries for: `libldap2-dev libsasl2-dev libssl-dev`
-- The latest [Python 3](https://www.python.org/) version and [pyenv](https://github.com/pyenv/pyenv)
+- The latest [Python 3](https://www.python.org/) version and
```
does uv necessarily need python to be installed?

jackie-ak replied:
Indeed, we don't need Python, just uv. And this is a step in the install.md anyways, so I removed the whole line here.

pmayer commented on env-skel:
```diff
@@ -1,7 +1,97 @@
+SITE_URL=https://url/
+
+## In a production deployment with other base components, this base path should
+## be set for image. For local development set this explicitly to an empty string.
```
is the "image" here a copy and paste error?

jackie-ak replied:
yep. fixed it and extended the info, that no adaptation is needed if a default prod setup is done.

pmayer commented on env-skel:
```diff
@@ -1,7 +1,97 @@
+# EMAIL_PORT=
+# EMAIL_USE_TLS=
+# EMAIL_USE_LOCALTIME=
+# EMAIL_SUBJECT_PREFIX=[CAS]
```
we should probably rename CAS to baseauth here

pmayer commented on src/api/views/autosuggest_user.py:
```diff
@@ -15,31 +17,33 @@
 def autosuggest_user(request, user, *args, **kwargs):
     """Get autosuggest results for query."""
-    UserModel = get_user_model()
+    User = get_user_model()  # noqa - this represents the model class
```
instead of just `noqa`, should we also mention the exact rule to ignore?

jackie-ak committed:
chore: apply migrations after linting changes on model fields

pmayer commented on docs/source/conf.py:
```diff
@@ -20,7 +20,7 @@
 
 project = 'baseauth'
-copyright = f'baseauth | University of Applied Arts Vienna, 2019-{date.today().year}'
+copyright = f'baseauth | University of Applied Arts Vienna, 2019-{date.today().year}'  # noqa - sphinx specific
```
should we add the exact rule here?

pmayer commented on src/api/views/autosuggest_user.py:
```diff
@@ -15,31 +17,33 @@
+            'UUID': user.username,  # TODO: discuss in review: should this be all caps? leaving for now, as frontend might depend on it
+            'first_name': user.first_name,
+            'last_name': user.last_name,
+            'label': f'{user.first_name} {user.last_name}',
```
the value of label should probably be `user.get_full_name()` instead of manually creating it.

pmayer commented on src/api/views/user_image.py:
```diff
@@ -53,16 +53,21 @@ def list(self, request, *args, **kwargs):
+                    ),
                 )
 
             return Response(
```
since we are touching this code, we could exchange the 404 responses to `NotFound`. i will only mention this once here and not for all 404 responses.

pmayer commented on src/api/views/user_preferences_agent.py:
```diff
@@ -38,12 +38,12 @@ def retrieve(self, request, pk=None, *args, **kwargs):
         except APIKey.DoesNotExist:
             raise PermissionDenied from None
-        UserModel = get_user_model()
+        User = get_user_model()  # noqa - this represents a model class
```
same as in the other file - should we mention the exact rule to ignore?

pmayer commented on src/api/views/user_preferences_data.py:
```diff
@@ -62,7 +60,7 @@ def retrieve(self, request, *args, **kwargs):
         },
     )
-    def _update(self, request, partial=False, *args, **kwargs):
+    def _update(self, request, partial=True, **kwargs):
```
is there a specific reason you changed the default value here?

jackie-ak replied:
No, I don't recall what the reason was. Actually I think `partial=False` is the safer/saner default.
But I tested it again, in the current implementation it is irrelevant which we set. This has to do with the `UserPreferencesDataSerializer` serializer where all three fields are configured as not required. Not sure why it was implemented in that way, but I think it would be better to make them required. Then the PATCH endpoint actually makes sense. What do you think?

jackie-ak replied:
Also checked, frontend currently only uses the PUT endpoint and always sends all three fields, even if they are an empty list or string.

jackie-ak replied:
Removed the PATCH endpoint, after confirming with frontend.

pmayer commented on src/api/views/user_preferences_data.py:
```diff
@@ -41,7 +39,7 @@ class UserPreferencesDataViewSet(GenericViewSet):
         },
     )
-    def retrieve(self, request, *args, **kwargs):
+    def retrieve(self, request, **kwargs):
```
shouldn't we keep `*args` here? also for the other functions in this file. is there a specific reason you removed it?

jackie-ak replied:
I think this was due to some ruff error, but because of the wrong ordering. I'll add it back to be consistent with the rest of the codebase. But generally I was also wondering why we do that in cases where we are definitely not using them? The drf docs also suggest not to use them per default. But maybe there is a good reason, other than IDE code completion? Should we discuss this in a backend meeting?

pmayer commented on src/api/views/user_settings.py:
```diff
@@ -17,14 +17,14 @@
 class UserSettingsViewSet(GenericViewSet, UpdateModelMixin):
     lookup_field = 'user'
-    '''
+    """
```
the comments here should probalby be removed or moved to the corresponding functions.

pmayer commented on src/api/views/user_settings.py:
```diff
@@ -76,19 +76,16 @@ def validate_value(self, value, value_type):
+        if isinstance(value, list) and (
+            not value or not all(isinstance(i, str) for i in value)
+        ):
+            # must be a list of strings
```
the comment should probably move up to the if clause.

pmayer commented on src/baseauth/settings.py:
```diff
@@ -218,14 +222,14 @@
 
 EMAIL_SUBJECT_PREFIX = '{} '.format(
-    env.str('EMAIL_SUBJECT_PREFIX', default='[CAS]').strip()
+    env.str('EMAIL_SUBJECT_PREFIX', default='[CAS]').strip(),
```
we should exchange CAS with baseauth here.

pmayer commented on src/baseauth/settings.py:
```diff
@@ -387,7 +392,7 @@
         'file': {
```
we could adapt the configuration to our new setup:

```
'file': {
    'level': 'DEBUG',
    'class': 'concurrent_log_handler.ConcurrentTimedRotatingFileHandler',
    'filename': LOG_DIR / 'application.log',
    'when': 'midnight',
    'backupCount': 1000,
    'use_gzip': True,
    'formatter': 'verbose',
},
```

pmayer commented on src/baseauth/urls.py:
```diff
@@ -126,7 +127,7 @@ def get_redirect_url(self, *args, **kwargs):
 if settings.DEBUG:
```
should we remove debug toolbar like we did in other projects?

pmayer commented on src/gunicorn_conf.py:
```diff
@@ -8,7 +8,7 @@
 timeout = 300
 workers = os.getenv('GUNICORN_WORKERS', multiprocessing.cpu_count() * 2 + 1)
-worker_tmp_dir = '/dev/shm'  # nosec
+worker_tmp_dir = '/dev/shm'  # noqa
```
we should add the explicit rule to ignore.

pmayer commented on src/requirements.in:
```diff
@@ -3,33 +3,35 @@
-django==4.2.7
-django-auth-ldap==4.6.0
-django-axes[ipware]==6.1.1
+django==4.2.21
```
should we use `django~=4.2` here like we do in other projects?

jackie-ak committed:
fix: use correct env vars for db container

jackie-ak committed:
fix: use baseauth for command instead of portfolio

jackie-ak committed:
docs: add installing pre-commit hooks to install.md

jackie-ak committed:
docs: add docker install step to prod install

jackie-ak committed:
fix: use baseauth in email subject prefix

jackie-ak committed:
change: use ConcurrentTimedRotatingFileHandler for log files

jackie-ak committed:
remove: drop debug_toolbar

jackie-ak committed:
chore: use explicit linter rule exception for gunicorn conf

jackie-ak committed:
build(deps): pin Django to latest 4.2

jackie-ak committed:
change: adapt default .env vars for postgres

jackie-ak committed:
docs: remove libssl-dev from requirement

This is already listed in our general local dev setup guide, so it
shouldn't be needed separately in this specific project.

jackie-ak commented:
Added two more commits. Now this should be ready for a merge, if nothing else turns up.

pmayer commented on .ruff.toml:
Is there a reason you copied the whole file from config? If not, i would keep it minimal as recommended in our [documentation](https://base-angewandte-docs.readthedocs.io/en/latest/dev_guide.html#coding-conventions):

```
extend = "config/.ruff.toml"

src = ["src"]
target-version = "py312"
```

jackie-ak replied:
Other than me not thinking? I guess not :blush: 
Fixed it now

pmayer commented on compose.yaml:
```diff
@@ -8,29 +7,39 @@ services:
-      - POSTGRES_PASSWORD=$BASEAUTH_DB_PASSWORD
-      - POSTGRES_USER=$BASEAUTH_DB_USER
-      - POSTGRES_DB=$BASEAUTH_DB_NAME
+      - POSTGRES_PASSWORD=$POSTGRES_PASSWORD
```
maybe a bit nit picky, but i would change the order to 

```
POSTGRES_DB
POSTGRES_USER
POSTBES_PASSWORD
```
so the order is the same as in the env file.

jackie-ak replied:
But makes sense. Adapted accordingly.

pmayer commented on docs/source/migration.md:
```diff
@@ -11,6 +11,22 @@ migrating from one of the versions listed below.
+**Config file changes:**
+
+We refactored the configuration to only use a single .env file. Also,
+the docker compose file was brought up to it's newest version. This
```
i think `it's` is a typo and should be `its`.

pmayer commented on env-skel:
```diff
@@ -1,7 +1,97 @@
+## PostgreSQL settings for baseauth-postgres container and Django
+POSTGRES_DB=django_baseauth
+POSTGRES_USER=django_baseauth
+POSTGRES_PASSWORD=
```
since we changed this variable to be empty, we should probably also remove the default value from [`settings.py`](https://github.com/base-angewandte/baseauth/blob/main/src/baseauth/settings.py#L262).

pmayer commented on src/user_preferences/management/commands/create_settings.py:
```diff
@@ -9,12 +9,14 @@ class Command(BaseCommand):
-            type=str,
-            help=_('A name for your app.'),
-        ),
+        (
```
this change looks strange, why are there new parentheses here?

jackie-ak replied:
No idea, why this was there. I guess this was somehow introduced when ruff went over the whole code base ... for whatever reason. Fixed it now.

pmayer reviewed (commented):
added a few more comments and noticed that there are still unresolved comments from the last review

jackie-ak committed:
change: extend instead of copy ruff settings

jackie-ak committed:
style: reorder env vars for consistency

jackie-ak committed:
docs: correct typo

jackie-ak committed:
change: remove default db pass from settings

jackie-ak committed:
change: remove enlcosing tuple from method call

jackie-ak committed:
docs: remove Python and uv from requirements.md

Only uv is needed (Python will be installed through uv), and this is
mentioned in the install.md, therefore removing it here.

jackie-ak committed:
docs: fix copy-paste text and extend info on FORCE_SCRIPT_NAME

jackie-ak committed:
docs: use correct default value in env-skel

jackie-ak committed:
style: make ruff exceptions more specific

jackie-ak committed:
change: use get_full_name() instead of manual creation

jackie-ak committed:
docs: use proper docstrings for api endpoints

jackie-ak committed:
style: improve comment style and content

jackie-ak committed:
refactor: use NotFound instead of manual 404 Response

jackie-ak committed:
docs: use proper docstrings for api endpoints

jackie-ak commented:
But during refactoring and debugging I also stumbled over the implementation [UserSettingsViewSet.update()](https://github.com/base-angewandte/baseauth/blob/fbe924c9757e9005bbe3d4ce32f68e268dc9cc44/src/api/views/user_settings.py#L88), which does not seem to be implemented properly. I was wondering if this endpoint is used anywhere. Also the general payload schema seems unnecessarily convoluted to me. Maybe we should discuss this separately?

jackie-ak committed:
remove: drop unused PATCH /user/data endpoint

jackie-ak commented:
Final remarks also resolved now, after confirming with frontend, that the PATCH endpoint for user data is not used.

pmayer commented on env-skel:
```diff
@@ -1,7 +1,98 @@
-## These settings are used to set up the DB in the basauth-postgres container.
-BASEAUTH_DB_NAME=django_baseauth
-BASEAUTH_DB_USER=django_baseauth
-## Make sure to change this to a strong password on any production/public server.
```
should we keep this comment? and add a general hint that a password has to be set?

jackie-ak replied:
Hm, we now explicitly document it in the configuration section of the docs, and the user would hit an error on startup, if they don't set it. But I guess an additional comment here would not hurt either. Something like
```
# A strong database password needs to be set explicitly here. Check the config docs for more.
```

pmayer commented on src/api/views/autosuggest_user.py:
```diff
@@ -15,31 +17,33 @@
-        )
+    response = [
+        {
+            'UUID': user.username,  # TODO: discuss in review: should this be all caps? leaving for now, as frontend might depend on it
```
i think, at the moment we are not consistent in our projects in how to send this information. but in image we use:

```
{
    'id': user.username,
    'label': user.get_full_name(),
},
```

so we probably should have the same structure. but that's definitely something that needs to be discussed with the FE dev team as well.

jackie-ak replied:
Yes, I personally would also tend to just use `id`. But let's discuss this and then adapt it within the api v2 branch. I created a separate ticket for that: https://github.com/base-angewandte/baseauth/issues/41

pmayer commented on src/api/views/user_settings.py:
```diff
@@ -75,31 +65,28 @@ def validate_value(self, value, value_type):
             return False
 
-        if not isinstance(value, value_type_mapping[value_type]):
+        if value_type not in value_type_mapping:
```
I think we should do this check first in the function, and maybe raise an exception instead of just returning `False`.

jackie-ak replied:
Well, the whole `validate_value` method in the ViewSet seems kind of misplaced. Why not use the serializer for that? I guess this is maybe an issue of legacy code, which we never really reviewed in cas?
But I would like to be able to close this PR before the long holiday, so I created a separate ticket for it, in which we could also discuss, whether and how to go about it: https://github.com/base-angewandte/baseauth/issues/40

pmayer reviewed (commented):
Some further comments.

jackie-ak committed:
docs: add comment on DB password in env-skel
